src/abstractdataframe/join.jl

```diff
@@ -100,16 +100,33 @@ check_mapping_allowed(short::AbstractVector, refarray_long::AbstractVector,
     !isempty(short) && !isnothing(refpool_long) && !isnothing(invrefpool_long) &&
         eltype(refarray_long) <: Union{Signed, Unsigned}
 
-@noinline map_refarray(mapping::AbstractVector, refarray::AbstractVector) =
+@noinline map_refarray(mapping::AbstractVector, refarray::AbstractVector, ::Val{true}) =
+    [@inbounds mapping[r + 1] for r in refarray]
+
+@noinline map_refarray(mapping::AbstractVector, refarray::AbstractVector, ::Val{false}) =
     [@inbounds mapping[r] for r in refarray]
 
-map2refs(x::AbstractVector, invrefpool) = [get(invrefpool, v, nothing) for v in x]
+function map2refs(x::AbstractVector, invrefpool)
+    x_refpool = DataAPI.refpool(x)
+    x_refarray = DataAPI.refarray(x)
+    if isnothing(x_refpool) || !(x_refpool isa AbstractVector) || !(eltype(x_refarray) <: Integer)
```

Comment:
It would be simpler to swap the two branches, and use this check to replace `if 0 <= fi <= 1` below:
```suggestion
    if x_refpool isa AbstractVector{<:Integer} && 0 <= firstindex(x_refpool) <= 1
```